NIX_USER_CONF_FILES

zimbatm opened the pull request:
add NIX_USER_CONF_FILES
    
Motivation: maintain project-level configuration files.
    
Document the whole situation a bit better so that it corresponds to the
implementation, and add NIX_USER_CONF_FILES that allows overriding
which user files Nix will load during startup.

domenkozar commented:
> The manual generation is currently failing. I tried understanding the error and bisecting the documentation and wasn't able to reduce the error to something actionable.

If this isn't a proof for https://github.com/NixOS/rfcs/pull/64 then I don't know what is :)

edolstra commented:
It's actually the opposite: DocBook at least has a concept of an "error". Markdown is garbage in, garbage out.

zimbatm commented:
I can fix the doc error with some more elbow grease. Can I get feedback on the PR instead? :D 

@edolstra does the PR make sense to you overall?

edolstra commented:
@zimbatm Yes, makes sense. However it may make more sense to add an environment variable `NIX_CONF_FILE` so you don't need to create a directory just for your configuration file. (E.g. a per-project config file could be in `<project-dir>/.nix.conf`).

edolstra commented:
Or generalize this to a *list* of configuration files (e.g. `NIX_CONF_FILES = my-project.conf /etc/nix/nix.conf ~/.config/nix/nix.conf `). After all, it may not be desirable to have the per-project config file completely override the user config file.

lheckemann commented:
What advantage does this have over using the existing XDG_CONF_DIRS support?

edolstra commented:
@lheckemann Good point, that might be the way to go.

roberth commented:
@lheckemann

> This is useful because sometimes it is desirable to be more selective than changing the config home of all XDG-respecting programs.

For example, you'll want your editor to read the usual configuration files.

lheckemann commented:
You don't need to change the config home (where programs write their config to), only add a dir to the config path (where they read their config from) — and if the extra entry in the config path only contains relevant config (i.e. nix config) it'll only affect nix.

andir commented:
On 06:52 01.04.20, Linus Heckemann wrote:
> You don't need to change the config home (where programs write their
> config to), only add a dir to the config path (where they read their
> config from) — and if the extra entry in the config path only contains
> relevant config (i.e. nix config) it'll only affect nix.

Yeah, I am just setting that as part of my .envrc:

> export XDG_CONFIG_DIRS="$(pwd)/.config:$XDG_CONFIG_DIRS"

It has served me well for a while now. If there was another variable on
the nix end it should probably just allow a list of files (as eelco
suggested).

roberth commented:
I was focused on `NIX_USER_CONF_DIR` so I didn't notice the plural in `XDG_CONFIG_DIRS`. This does seem adequate for most cases. If you don't want an extra ".config" directory in your project, it seems like you could generate it in a `shell.nix` dependency.

zimbatm commented:
@lheckemann `XDG_CONFIG_DIRS` is loaded before `XDG_CONFIG_HOME` as per the XDG spec. If the user has set any configuration they will take precedence over XDG_CONFIG_DIRS.

zimbatm commented:
A variation on the XDG idea would be to push `XDG_CONFIG_DIRS=$XDG_CACHE_HOME:$XDG_CONFIG_DIRS` and then override `XDG_CACHE_HOME=$PWD/config` where `$PWD` is the project root. Unfortunately, the first thing I noticed is that `git` only implements the XDG spec partially and doesn't look at `XDG_CONFIG_DIRS`. So I think this PR is really needed.

zimbatm commented on doc/manual/command-ref/conf-file.xml:
```diff
@@ -29,13 +29,35 @@
+      <filename>$NIX_USER_CONF_DIR/nix.conf</filename>.
+      where <envvar>NIX_USER_CONF_DIR</envvar> defaults to
+      <filename>$XDG_CONFIG_HOME/nix</filename>
+      and <envvar>XDG_CONFIG_HOME</envvar> defaults to
```
I found the XML error:
```suggestion
      and <envar>XDG_CONFIG_HOME</envar> defaults to
```

zimbatm commented:
I fixed the doc generation and added a minimal test. The last thing to address is @edolstra's comment on [`NIX_CONF_FILE` and `NIX_CONF_FILES`.](https://github.com/NixOS/nix/pull/3458#issuecomment-607181594).

zimbatm commented:
@edolstra the env has been renamed to NIX_CONF_FILES and now appends to the existing config paths.

matthewbauer commented:
Does this override the system configuration or just append to it? For instance, what happens when a local config has `substituters = ...`. Is https://cache.nixos.org/ still loaded?

zimbatm commented:
Keys get overridden and the last writer wins. For example, if `/etc/nix/nix.conf` sets the `substituters` key and `~/.config/nix/nix.conf` as well, then the user config will take precedence. It's a valid concern but I think that should be handled in a separate PR as it extends the scope of the work and is not a new issue introduced by this PR.

domenkozar commented:
That's https://github.com/NixOS/nix/issues/2149

zimbatm commented:
Who is able to give this PR another substantial review? Does it make sense how I changed the code structure to load all the config files?

PkmX commented on doc/manual/command-ref/conf-file.xml:
```diff
@@ -29,13 +29,36 @@
+  <listitem>
+    <para>
+      And finally, every files pointed to by
+      <envar>NIX_EXTRA_CONF_FILES</envar>, separated by <literal>:</literal>
```
Is this implemented?

zimbatm replied:
yes? You can see it in action in the admittedly minimal test.

PkmX replied:
It only uses `NIX_CONF_FILES`, not `NIX_EXTRA_CONF_FILES`.

zimbatm replied:
oh ok. This is a typo. Fixed in 976ead5229d776a9559629b279bd7a27f9945111

PkmX commented:
It also seems like the `NIX_CONF_FILES` as implemented are actually extra conf files that are read after `$XDG_CONFIG_DIRS` and `$XDG_CONFIG_HOME`, but I think @edolstra's comment about `NIX_CONF_FILES` is that it specifies the absolute list of conf files to be loaded ignoring `$NIX_CONF_DIR/nix.conf` and all users config files. This would fit my use case where my build system is essentially a wrapper over Nix and needs exact control over the config, and I don't want any system-wide or user-wide `nix.conf` to be loaded.

edolstra commented on doc/manual/command-ref/conf-file.xml:
```diff
@@ -19,7 +19,7 @@
 <refsection><title>Description</title>
 
-<para>Nix reads settings from two configuration files:</para>
+<para>Nix reads settings from the following places, in that order:</para>
```
'in this order'

zimbatm replied:
fixed!

edolstra commented on doc/manual/command-ref/conf-file.xml:
```diff
@@ -29,13 +29,36 @@
     <filename>$NIX_CONF_DIR/nix.conf</filename> if
     <envar>NIX_CONF_DIR</envar> is set.</para>
+
+    <para>Values loaded in that file are not forwarded to the nix-daemon. The
```
'in this file'

'Nix daemon'

edolstra commented on doc/manual/command-ref/conf-file.xml:
```diff
@@ -29,13 +29,36 @@
+
+  <listitem>
+    <para>
+      For each folder in <envar>XDG_CONFIG_DIRS</envar>, separated by
```
'each directory'

edolstra commented on doc/manual/command-ref/conf-file.xml:
```diff
@@ -29,13 +29,36 @@
+      For each folder in <envar>XDG_CONFIG_DIRS</envar>, separated by
+      <literal>:</literal>.
+
+      Load <filename><replaceable>folder</replaceable>/nix/nix.conf</filename>.
```
'directory'

edolstra commented on doc/manual/command-ref/conf-file.xml:
```diff
@@ -29,13 +29,36 @@
+
+  <listitem>
+    <para>
+      And finally, every files pointed to by
```
'every file'

edolstra commented on doc/manual/command-ref/conf-file.xml:
```diff
@@ -29,13 +29,36 @@
+    <para>
+      And finally, every files pointed to by
+      <envar>NIX_CONF_FILES</envar>, separated by <literal>:</literal>
+      if that environment variable exists.
```
This is phrased awkwardly (it's the *contents* of NIX_CONF_FILE that's separated by `:`).

edolstra commented on doc/manual/command-ref/env-common.xml:
```diff
@@ -137,12 +137,19 @@ $ mount -o bind /mnt/otherdisk/nix /nix</screen>
 
+<varlistentry><term><envar>NIX_CONF_FILES</envar></term>
+
+  <listitem><para>Specifies another set of nix configuration files to load
```
`Nix`

edolstra commented on doc/manual/command-ref/env-common.xml:
```diff
@@ -137,12 +137,19 @@ $ mount -o bind /mnt/otherdisk/nix /nix</screen>
+<varlistentry><term><envar>NIX_CONF_FILES</envar></term>
+
+  <listitem><para>Specifies another set of nix configuration files to load
+  that override the default locations. Each file is seprated by
```
'separated'

edolstra commented on src/libstore/globals.cc:
```diff
@@ -77,11 +78,26 @@ void loadConfFile()
-    // Iterate over them in reverse so that the ones appearing first in the path take priority
-    for (auto dir = dirs.rbegin(); dir != dirs.rend(); dir++) {
-        globalConfig.applyConfigFile(*dir + "/nix/nix.conf");
+    for (auto dir = dirs.begin(); dir != dirs.end(); dir++) {
```
`for (auto & dir : dirs)`

zimbatm replied:
much nicer

edolstra commented on src/libutil/util.hh:
```diff
@@ -387,6 +388,24 @@ string replaceStrings(const std::string & s,
 
+/* Create a string from a Range (like std::vector), separated by "delimited" */
+template <typename Range, typename Value = typename Range::value_type>
+std::string join(Range const& elements, const char *const delimiter) {
```
This looks a lot like the `concatStringsSep` function in util.cc.

zimbatm replied:
Indeed! Removed

zimbatm commented:
This is ready for another round of review.

* I changed `NIX_CONF_FILES` to `NIX_USER_CONF_FILES` to be more specific and also indicate that it overrides the user configuration. 
* I changed the logic so that `NIX_USER_CONF_FILES` replaces the default locations where user config is being loaded. The system `/etc/nix/nix.conf` is still being loaded in any case as it plays a special role.
* Fixed the docs, probably introduced some more language issues
